File: backend/api/routers/git.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from backend.api.dependencies import get_container
from backend.api.schemas.git import (
    AuditReference,
    CheckoutRefRequest,
    CloneRepositoryRequest,
    CreateBranchRequest,
    CreateCommitRequest,
    DeleteBranchRequest,
    DiscoverGitRequest,
    ErrorPayload,
    PullRepositoryRequest,
    ResponseEnvelope,
)
from backend.application.commands import CommandExecutionResult, CommandPreview
from backend.application.git import GitApplicationError
from backend.domain.shared_kernel import ProjectId
from backend.infrastructure.di import ApplicationContainer


router = APIRouter(prefix="/api/v1", tags=["git"])


@router.get("/projects/{project_id}/git/repositories", response_model=ResponseEnvelope)
def list_repositories(project_id: str, role: str | None = None, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    return _success(container.create_git_service().list_git_repositories(ProjectId(project_id), role))


@router.get("/projects/{project_id}/git/repositories/{repo_id}", response_model=ResponseEnvelope)
def get_repository(project_id: str, repo_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _success(container.create_git_service().get_git_repository(ProjectId(project_id), repo_id))
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/discover", response_model=ResponseEnvelope)
def discover_repositories(project_id: str, request: DiscoverGitRequest, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _success(container.create_git_service().execute_discover_git_repositories(project_id=ProjectId(project_id), path_filters=request.path_filters))
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/clone-plan", response_model=ResponseEnvelope)
def plan_clone(project_id: str, request: CloneRepositoryRequest, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    preview = container.create_git_service().preview_clone_repository(
        project_id=ProjectId(project_id),
        remote_url=request.remote_url,
        destination_path=request.destination_path,
        repository_role=request.repository_role,
    )
    return _success(_preview_data(preview), warnings=preview.warnings)


@router.post("/projects/{project_id}/git/clone", response_model=ResponseEnvelope)
def clone_repository(project_id: str, request: CloneRepositoryRequest, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _command_success(
            container.create_git_service().execute_clone_repository(
                project_id=ProjectId(project_id),
                remote_url=request.remote_url,
                destination_path=request.destination_path,
                repository_role=request.repository_role,
            )
        )
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/repositories/{repo_id}/fetch-plan", response_model=ResponseEnvelope)
def plan_fetch(project_id: str, repo_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        preview = container.create_git_service().preview_fetch_repository(project_id=ProjectId(project_id), git_repository_id=repo_id)
        return _success(_preview_data(preview), warnings=preview.warnings)
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/repositories/{repo_id}/fetch", response_model=ResponseEnvelope)
def fetch_repository(project_id: str, repo_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _command_success(container.create_git_service().execute_fetch_repository(project_id=ProjectId(project_id), git_repository_id=repo_id))
    except GitApplicationError as error:
        return _failure(error)


@router.get("/projects/{project_id}/git/repositories/{repo_id}/status", response_model=ResponseEnvelope)
def get_status(project_id: str, repo_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _success(container.create_git_service().get_worktree_status(ProjectId(project_id), repo_id))
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/repositories/{repo_id}/pull-plan", response_model=ResponseEnvelope)
def plan_pull(project_id: str, repo_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        preview = container.create_git_service().preview_pull_repository(project_id=ProjectId(project_id), git_repository_id=repo_id)
        return _success(_preview_data(preview), warnings=preview.warnings)
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/repositories/{repo_id}/pull", response_model=ResponseEnvelope)
def pull_repository(project_id: str, repo_id: str, request: PullRepositoryRequest, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _command_success(
            container.create_git_service().execute_pull_repository(
                project_id=ProjectId(project_id),
                git_repository_id=repo_id,
                idempotency_key=request.idempotency_key,
            )
        )
    except GitApplicationError as error:
        return _failure(error)


@router.get("/projects/{project_id}/git/repositories/{repo_id}/refs", response_model=ResponseEnvelope)
def list_refs(project_id: str, repo_id: str, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _success(container.create_git_service().list_refs(ProjectId(project_id), repo_id))
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/repositories/{repo_id}/branches", response_model=ResponseEnvelope)
def create_branch(project_id: str, repo_id: str, request: CreateBranchRequest, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _command_success(
            container.create_git_service().execute_create_branch(
                project_id=ProjectId(project_id),
                git_repository_id=repo_id,
                branch_name=request.branch_name,
                idempotency_key=request.idempotency_key,
            )
        )
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/repositories/{repo_id}/checkout", response_model=ResponseEnvelope)
def checkout_ref(project_id: str, repo_id: str, request: CheckoutRefRequest, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _command_success(
            container.create_git_service().execute_checkout_ref(
                project_id=ProjectId(project_id),
                git_repository_id=repo_id,
                ref_name=request.ref_name,
                idempotency_key=request.idempotency_key,
            )
        )
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/repositories/{repo_id}/branches/delete-plan", response_model=ResponseEnvelope)
def plan_delete_branch(project_id: str, repo_id: str, request: DeleteBranchRequest, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        preview = container.create_git_service().preview_delete_branch(
            project_id=ProjectId(project_id), git_repository_id=repo_id, branch_name=request.branch_name
        )
        return _success(_preview_data(preview), warnings=preview.warnings)
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/repositories/{repo_id}/branches/delete", response_model=ResponseEnvelope)
def delete_branch(project_id: str, repo_id: str, request: DeleteBranchRequest, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _command_success(
            container.create_git_service().execute_delete_branch(
                project_id=ProjectId(project_id),
                git_repository_id=repo_id,
                branch_name=request.branch_name,
                idempotency_key=request.idempotency_key,
            )
        )
    except GitApplicationError as error:
        return _failure(error)


@router.get("/projects/{project_id}/git/repositories/{repo_id}/diff", response_model=ResponseEnvelope)
def get_diff(
    project_id: str,
    repo_id: str,
    base_ref: str = Query(...),
    head_ref: str = Query(...),
    path_filter: str | None = None,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(container.create_git_service().get_diff_summary(ProjectId(project_id), repo_id, base_ref, head_ref, path_filter))
    except GitApplicationError as error:
        return _failure(error)


@router.get("/projects/{project_id}/git/repositories/{repo_id}/commits/compare", response_model=ResponseEnvelope)
def compare_commits(
    project_id: str,
    repo_id: str,
    base_ref: str = Query(...),
    head_ref: str = Query(...),
    limit: int = 20,
    container: ApplicationContainer = Depends(get_container),
) -> ResponseEnvelope:
    try:
        return _success(container.create_git_service().compare_commits(ProjectId(project_id), repo_id, base_ref, head_ref, limit))
    except GitApplicationError as error:
        return _failure(error)


@router.post("/projects/{project_id}/git/repositories/{repo_id}/commits", response_model=ResponseEnvelope)
def create_commit(project_id: str, repo_id: str, request: CreateCommitRequest, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    try:
        return _command_success(
            container.create_git_service().execute_create_commit(
                project_id=ProjectId(project_id),
                git_repository_id=repo_id,
                message=request.message,
                paths=request.paths,
                idempotency_key=request.idempotency_key,
            )
        )
    except GitApplicationError as error:
        return _failure(error)


@router.get("/projects/{project_id}/git/operations", response_model=ResponseEnvelope)
def list_operations(project_id: str, repo_id: str | None = None, container: ApplicationContainer = Depends(get_container)) -> ResponseEnvelope:
    return _success(container.create_git_service().list_git_operations(ProjectId(project_id), repo_id))


def _success(data: dict | list[dict], warnings: list[str] | None = None) -> ResponseEnvelope:
    return ResponseEnvelope(ok=True, data=data, warnings=warnings or [])


def _failure(error: GitApplicationError) -> ResponseEnvelope:
    return ResponseEnvelope(ok=False, error=ErrorPayload(code=error.code.value, message=str(error)))


def _command_success(result: CommandExecutionResult) -> ResponseEnvelope:
    return ResponseEnvelope(
        ok=True,
        data={
            **result.result,
            "command_plan_id": str(result.command_plan_id),
            "command_execution_id": str(result.command_execution_id),
            "undo_plan": result.undo_plan.payload if result.undo_plan else None,
        },
        audit_ref=AuditReference(ref_type=result.audit_ref.ref_type, ref_id=result.audit_ref.ref_id),
    )


def _preview_data(preview: CommandPreview) -> dict:
    return {"command_type": preview.command_type, "summary": preview.summary, "risk_level": preview.risk_level.value, "preview": preview.preview}

File: frontend/src/api/git.ts
import { jsonRequest, requestBackend, type BackendResponse } from "./backend";
import type { CommandPreviewData, CommandResultData, DryRunData } from "./project";

export interface GitRepository {
  git_repository_id: string;
  project_id: string;
  local_path: string;
  remote_url: string | null;
  default_branch: string | null;
  repository_role: string;
  last_scanned_at: string | null;
}

export interface GitRef {
  ref_name: string;
  ref_type: string;
  commit_sha: string;
  is_current: boolean;
}

export interface GitFileChange {
  path: string;
  change_status: string;
  old_path: string | null;
  insertions: number;
  deletions: number;
}

export interface GitStatus {
  head_commit_sha: string;
  branch_name: string | null;
  is_dirty: boolean;
  ahead_count: number;
  behind_count: number;
  file_changes: GitFileChange[];
  summary: string;
}

export interface GitCommit {
  commit_sha: string;
  parent_shas: string[];
  author_name: string;
  committed_at: string;
  message_summary: string;
}

export interface CloneRepositoryRequest {
  remote_url: string;
  destination_path: string;
  repository_role?: string;
}

export type GitCommandResponse = BackendResponse<CommandResultData>;

export async function listGitRepositories(projectId: string, role?: string | null): Promise<GitRepository[]> {
  const query = role ? `?role=${encodeURIComponent(role)}` : "";
  return (await requestBackend<GitRepository[]>(`/api/v1/projects/${projectId}/git/repositories${query}`)).data;
}

export async function getGitRepository(projectId: string, repoId: string): Promise<GitRepository> {
  return (await requestBackend<GitRepository>(`/api/v1/projects/${projectId}/git/repositories/${repoId}`)).data;
}

export async function discoverGitRepositories(projectId: string, pathFilters?: string[] | null): Promise<Record<string, unknown>> {
  return (await requestBackend<Record<string, unknown>>(`/api/v1/projects/${projectId}/git/discover`, jsonRequest({ path_filters: pathFilters ?? null }))).data;
}

export async function previewCloneRepository(projectId: string, request: CloneRepositoryRequest): Promise<BackendResponse<CommandPreviewData>> {
  return requestBackend<CommandPreviewData>(`/api/v1/projects/${projectId}/git/clone-plan`, jsonRequest(request));
}

export async function dryRunCloneRepository(projectId: string, request: CloneRepositoryRequest): Promise<BackendResponse<DryRunData>> {
  return requestBackend<DryRunData>(`/api/v1/projects/${projectId}/git/clone-dry-run`, jsonRequest(request));
}

export async function cloneRepository(projectId: string, request: CloneRepositoryRequest): Promise<GitCommandResponse> {
  return requestBackend<CommandResultData>(`/api/v1/projects/${projectId}/git/clone`, jsonRequest(request));
}

export async function fetchRepository(projectId: string, repoId: string): Promise<GitCommandResponse> {
  return requestBackend<CommandResultData>(`/api/v1/projects/${projectId}/git/repositories/${repoId}/fetch`, jsonRequest({}));
}

export async function previewPullRepository(projectId: string, repoId: string): Promise<BackendResponse<CommandPreviewData>> {
  return requestBackend<CommandPreviewData>(`/api/v1/projects/${projectId}/git/repositories/${repoId}/pull-plan`, jsonRequest({}));
}

export async function dryRunPullRepository(projectId: string, repoId: string): Promise<BackendResponse<DryRunData>> {
  return requestBackend<DryRunData>(`/api/v1/projects/${projectId}/git/repositories/${repoId}/pull-dry-run`, jsonRequest({}));
}

export async function pullRepository(projectId: string, repoId: string, idempotencyKey?: string | null): Promise<GitCommandResponse> {
  return requestBackend<CommandResultData>(
    `/api/v1/projects/${projectId}/git/repositories/${repoId}/pull`,
    jsonRequest({ idempotency_key: idempotencyKey ?? crypto.randomUUID() })
  );
}

export async function getGitStatus(projectId: string, repoId: string): Promise<GitStatus> {
  return (await requestBackend<GitStatus>(`/api/v1/projects/${projectId}/git/repositories/${repoId}/status`)).data;
}

export async function listGitRefs(projectId: string, repoId: string): Promise<GitRef[]> {
  return (await requestBackend<GitRef[]>(`/api/v1/projects/${projectId}/git/repositories/${repoId}/refs`)).data;
}

export async function createBranch(projectId: string, repoId: string, branchName: string): Promise<GitCommandResponse> {
  return requestBackend<CommandResultData>(
    `/api/v1/projects/${projectId}/git/repositories/${repoId}/branches`,
    jsonRequest({ branch_name: branchName, idempotency_key: crypto.randomUUID() })
  );
}

export async function checkoutRef(projectId: string, repoId: string, refName: string): Promise<GitCommandResponse> {
  return requestBackend<CommandResultData>(
    `/api/v1/projects/${projectId}/git/repositories/${repoId}/checkout`,
    jsonRequest({ ref_name: refName, idempotency_key: crypto.randomUUID() })
  );
}

export async function previewDeleteBranch(projectId: string, repoId: string, branchName: string): Promise<BackendResponse<CommandPreviewData>> {
  return requestBackend<CommandPreviewData>(
    `/api/v1/projects/${projectId}/git/repositories/${repoId}/branches/delete-plan`,
    jsonRequest({ branch_name: branchName })
  );
}

export async function dryRunDeleteBranch(projectId: string, repoId: string, branchName: string): Promise<BackendResponse<DryRunData>> {
  const preview = await previewDeleteBranch(projectId, repoId, branchName);
  return {
    data: {
      command_type: preview.data.command_type,
      valid: true,
      simulation: preview.data.preview
    },
    warnings: preview.warnings,
    auditRef: preview.auditRef
  };
}

export async function deleteBranch(projectId: string, repoId: string, branchName: string): Promise<GitCommandResponse> {
  return requestBackend<CommandResultData>(
    `/api/v1/projects/${projectId}/git/repositories/${repoId}/branches/delete`,
    jsonRequest({ branch_name: branchName, idempotency_key: crypto.randomUUID() })
  );
}

export async function getGitDiff(
  projectId: string,
  repoId: string,
  baseRef: string,
  headRef: string,
  pathFilter?: string | null
): Promise<Record<string, unknown>> {
  const params = new URLSearchParams({ base_ref: baseRef, head_ref: headRef });
  if (pathFilter) {
    params.set("path_filter", pathFilter);
  }
  return (await requestBackend<Record<string, unknown>>(`/api/v1/projects/${projectId}/git/repositories/${repoId}/diff?${params}`)).data;
}

export async function compareCommits(
  projectId: string,
  repoId: string,
  baseRef: string,
  headRef: string,
  limit = 20
): Promise<GitCommit[]> {
  const params = new URLSearchParams({ base_ref: baseRef, head_ref: headRef, limit: String(limit) });
  return (await requestBackend<GitCommit[]>(`/api/v1/projects/${projectId}/git/repositories/${repoId}/commits/compare?${params}`)).data;
}

export async function createCommit(
  projectId: string,
  repoId: string,
  message: string,
  paths?: string[] | null
): Promise<GitCommandResponse> {
  return requestBackend<CommandResultData>(
    `/api/v1/projects/${projectId}/git/repositories/${repoId}/commits`,
    jsonRequest({ message, paths: paths ?? null, idempotency_key: crypto.randomUUID() })
  );
}

export async function listGitOperations(projectId: string, repoId?: string | null): Promise<Record<string, unknown>[]> {
  const query = repoId ? `?repo_id=${encodeURIComponent(repoId)}` : "";
  return (await requestBackend<Record<string, unknown>[]>(`/api/v1/projects/${projectId}/git/operations${query}`)).data;
}
